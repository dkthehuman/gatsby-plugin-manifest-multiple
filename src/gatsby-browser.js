import ManifestHelper from './ManifestHelper';

function hasDiffAttr(el1, el2, attr) {
  return el1 && el2 && el1.getAttribute(attr) !== el2.getAttribute(attr);
}

function promptListener(e) {
  // Prevent Chrome 76 and later from showing the mini-infobar
  e.preventDefault();
}

exports.onRouteUpdate = function({ location }, pluginOptions) {
  const { childApps } = pluginOptions;
  if (!childApps) {
    return;
  }

  const helper = new ManifestHelper(
    location.pathname,
    pluginOptions,
    ManifestHelper.TAG_TYPES.HTML
  );

  const oldTags = [];
  const newTags = [];

  const currentManifestEl = document.head.querySelector('link[rel="manifest"]');
  const manifestEl = helper.getManifestLinkTag();
  if (hasDiffAttr(currentManifestEl, manifestEl, 'href')) {
    oldTags.push(currentManifestEl);
    newTags.push(manifestEl);
  }

  const currentFaviconEl = document.head.querySelector(
    'link[rel="shortcut icon"]'
  );
  const faviconEl = helper.getFaviconLinkTag();
  if (hasDiffAttr(currentFaviconEl, faviconEl, 'href')) {
    oldTags.push(currentFaviconEl);
    newTags.push(faviconEl);
  }

  const currentThemeColorEl = document.head.querySelector(
    'meta[name="theme-color"]'
  );
  const themeColorEl = helper.getThemeColorMetaTag();
  if (hasDiffAttr(currentThemeColorEl, themeColorEl, 'content')) {
    oldTags.push(currentThemeColorEl);
    newTags.push(themeColorEl);
  }

  const currentAppleIconEls = document.head.querySelectorAll(
    'link[rel="apple-touch-icon"]'
  );
  const appleIconEls = helper.getAppleIconLinkTags();
  const updateAppleIcons =
    currentAppleIconEls.length > 0 &&
    appleIconEls.length > 0 &&
    hasDiffAttr(currentAppleIconEls[0], appleIconEls[0], 'href');
  if (updateAppleIcons) {
    oldTags.push(...currentAppleIconEls);
    newTags.push(...appleIconEls);
  }

  oldTags.forEach(tag => tag.parentNode.removeChild(tag));
  newTags.forEach(tag => document.head.appendChild(tag));

  // Note that 'beforeinstallprompt' is only called on initial page load and not on subsequent navigations
  window.removeEventListener('beforeinstallprompt', promptListener);
  if (helper.shouldHideChromeMiniBar()) {
    window.addEventListener('beforeinstallprompt', promptListener);
  }
};
