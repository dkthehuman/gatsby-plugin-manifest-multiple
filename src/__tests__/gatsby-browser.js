import { onRouteUpdate } from '../gatsby-browser';

const pluginOptions = {
  name: 'My Website',
  start_url: '/',
  icon: 'main-icon.png',
  childApps: [
    {
      start_url: '/app/',
      name: 'My App',
      icon: 'child-icon.png',
    },
  ],
};

beforeAll(() => {
  global.__PATH_PREFIX__ = '';
});

describe('Manifest', () => {
  beforeEach(() => {
    document.head.innerHTML =
      '<link rel="manifest" href="/manifest.webmanifest">';
  });

  test('updates manifest if navigating to child app', () => {
    const location = {
      pathname: '/app/',
    };
    onRouteUpdate({ location }, pluginOptions);
    expect(document.head).toMatchInlineSnapshot(`
      <head>
        <link
          href="/manifest_my-app.webmanifest"
          rel="manifest"
        />
      </head>
    `);
  });

  test('keeps default manifest if not navigating to child app', () => {
    const location = {
      pathname: '/random-path/',
    };
    onRouteUpdate({ location }, pluginOptions);
    expect(document.head).toMatchInlineSnapshot(`
      <head>
        <link
          href="/manifest.webmanifest"
          rel="manifest"
        />
      </head>
    `);
  });
});

describe('Apple Icons', () => {
  beforeEach(() => {
    document.head.innerHTML =
      '<link rel="apple-touch-icon" sizes="48x48" href="/icons/icon-48x48.png">';
  });

  test('updates apple icons if navigating to child app', () => {
    const location = {
      pathname: '/app/',
    };
    onRouteUpdate({ location }, pluginOptions);
    expect(document.head).toMatchSnapshot();
  });

  test(`keeps apple icons if child app doesn't have own icon`, () => {
    let options = {
      name: 'My Website',
      start_url: '/',
      icon: 'main-icon.png',
      childApps: [
        {
          start_url: '/app/',
          name: 'My App',
        },
      ],
    };
    const location = {
      pathname: '/app/',
    };
    onRouteUpdate({ location }, options);
    expect(document.head).toMatchInlineSnapshot(`
      <head>
        <link
          href="/icons/icon-48x48.png"
          rel="apple-touch-icon"
          sizes="48x48"
        />
      </head>
    `);
  });

  test('keeps apple icons if not navigating to child app', () => {
    const location = {
      pathname: '/random-path/',
    };
    onRouteUpdate({ location }, pluginOptions);
    expect(document.head).toMatchInlineSnapshot(`
      <head>
        <link
          href="/icons/icon-48x48.png"
          rel="apple-touch-icon"
          sizes="48x48"
        />
      </head>
    `);
  });

  describe('Mini bar', () => {
    const promptEvent = {
      preventDefault: jest.fn(),
    };
    window.addEventListener = jest.fn((type, callback) => {
      callback(promptEvent);
    });

    beforeEach(() => {
      window.addEventListener.mockClear();
      promptEvent.preventDefault.mockClear();
    });

    test('shows mini-bar by default', () => {
      const location = {
        pathname: '/',
      };
      onRouteUpdate({ location }, pluginOptions);
      expect(promptEvent.preventDefault).not.toHaveBeenCalled();
    });

    test('prevents mini-bar when set to hidden', () => {
      const testOptions = {
        ...pluginOptions,
        ...{
          show_chrome_mini_bar: false,
        },
      };

      const location = {
        pathname: '/',
      };
      onRouteUpdate({ location }, testOptions);
      expect(promptEvent.preventDefault).toHaveBeenCalled();
    });

    test('shows mini-bar when hidden for parent but shown for child', () => {
      const testOptions = {
        ...pluginOptions,
        ...{
          show_chrome_mini_bar: false,
        },
      };
      testOptions.childApps[0].show_chrome_mini_bar = true;

      onRouteUpdate({ location: { pathname: '/' } }, testOptions);
      expect(promptEvent.preventDefault).toHaveBeenCalled();

      promptEvent.preventDefault.mockClear();
      onRouteUpdate({ location: { pathname: '/app/' } }, testOptions);
      expect(promptEvent.preventDefault).not.toHaveBeenCalled();
    });
  });
});
