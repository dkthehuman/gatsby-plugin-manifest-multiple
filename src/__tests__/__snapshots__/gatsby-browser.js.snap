// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Apple Icons updates apple icons if navigating to child app 1`] = `
<head>
  <link
    href="/icons/icon_my-app-48x48.png"
    rel="apple-touch-icon"
    sizes="48x48"
  />
  <link
    href="/icons/icon_my-app-72x72.png"
    rel="apple-touch-icon"
    sizes="72x72"
  />
  <link
    href="/icons/icon_my-app-96x96.png"
    rel="apple-touch-icon"
    sizes="96x96"
  />
  <link
    href="/icons/icon_my-app-144x144.png"
    rel="apple-touch-icon"
    sizes="144x144"
  />
  <link
    href="/icons/icon_my-app-192x192.png"
    rel="apple-touch-icon"
    sizes="192x192"
  />
  <link
    href="/icons/icon_my-app-256x256.png"
    rel="apple-touch-icon"
    sizes="256x256"
  />
  <link
    href="/icons/icon_my-app-384x384.png"
    rel="apple-touch-icon"
    sizes="384x384"
  />
  <link
    href="/icons/icon_my-app-512x512.png"
    rel="apple-touch-icon"
    sizes="512x512"
  />
</head>
`;
